torch/csrc/jit/codegen/cuda/runtime/welford.cu

```diff
@@ -0,0 +1,423 @@
+    const _dim3bd& block_dim,
+    T* shared_mem_M2,
+    T* shared_mem_avg,
+    bool read_write_pred,
+    T init_val) {
+  unsigned int reduction_size = (X_REDUCE ? block_dim.x : 1) *
+      (Y_REDUCE ? block_dim.y : 1) * (Z_REDUCE ? block_dim.z : 1);
+  // If this thread will output a final result
+  bool should_write = true;
+  if (X_REDUCE)
+    should_write = should_write && thread_idx.x == 0;
+  if (Y_REDUCE)
+    should_write = should_write && thread_idx.y == 0;
+  if (Z_REDUCE)
+    should_write = should_write && thread_idx.z == 0;
+  unsigned int reduction_stride;
+  unsigned int reduction_tid;
+  unsigned int linear_tid;
+  if (X_REDUCE && !Y_REDUCE && Z_REDUCE) {
+    // Transpose Z and Y in the shared memory so Z and X dims are contiguous in
+    // smem
+    reduction_stride = 1;
+    linear_tid = threadIdx.y * blockDim.z * blockDim.x +
+        threadIdx.z * blockDim.x + threadIdx.x;
+    reduction_tid = threadIdx.z * blockDim.x + threadIdx.x;
+  } else {
+    // Normal reduction in order
+    reduction_stride =
+        (X_REDUCE ? 1
+                  : (Y_REDUCE ? block_dim.x
+                              : (Z_REDUCE ? block_dim.x * block_dim.y : 0)));
+    linear_tid = thread_idx.z * block_dim.y * block_dim.x +
+        thread_idx.y * block_dim.x + thread_idx.x;
+    reduction_tid = (Z_REDUCE ? thread_idx.z : 0) *
+            (Y_REDUCE ? block_dim.y : 1) * (X_REDUCE ? block_dim.x : 1) +
+        (Y_REDUCE ? thread_idx.y : 0) * (X_REDUCE ? block_dim.x : 1) +
+        (X_REDUCE ? thread_idx.x : 0);
+  }
+  // assert( reduction_stride != 0 );
+  if (read_write_pred) {
+    shared_mem_M2[linear_tid] = in_M2;
+    shared_mem_avg[linear_tid] = in_avg;
+  } else {
+    shared_mem_M2[linear_tid] = init_val;
+    shared_mem_avg[linear_tid] = init_val;
+  }
+  __syncthreads();
+  // Reduce down to nearest power of 2:
+  int np2 = 1 << (31 - __clz(reduction_size));
+  int lower = N;
+  int upper = 0;
+  int border = reduction_size;
+  if (reduction_tid < np2) {
+    if (reduction_tid + np2 < reduction_size) {
+      int tmp_ = getCount(lower, upper, border, reduction_tid);
+      welfordCombine(
+          shared_mem_M2[linear_tid],
+          shared_mem_avg[linear_tid],
+          tmp_,
+          shared_mem_M2[linear_tid + np2 * reduction_stride],
+          shared_mem_avg[linear_tid + np2 * reduction_stride],
+          getCount(lower, upper, border, reduction_tid + np2));
+    }
+    foldCount(lower, upper, border, np2);
+  }
+  __syncthreads();
+  for (int factor = np2 / 2; factor > 0; factor >>= 1) {
+    if (reduction_tid < factor) {
+      int tmp_ = getCount(lower, upper, border, reduction_tid);
+      welfordCombine(
+          shared_mem_M2[linear_tid],
+          shared_mem_avg[linear_tid],
+          tmp_,
+          shared_mem_M2[linear_tid + factor * reduction_stride],
+          shared_mem_avg[linear_tid + factor * reduction_stride],
+          getCount(lower, upper, border, reduction_tid + factor));
```

Comment:
Ditto: count of B should also be accumulated during tree reduction.

Alternative than computing this, dumping them in shared memory is always a safe bet.

Reply:
Same as the above thread. The actual count accumulation happens in `foldCount` function. Although `welfordCombine` does the accumulation as well, it was intended for serial version of welford and the final combine with initial count. Block parallel version is not using that value.